### PR TITLE
[HTTPDB] Retry on POST for idempotent paths

### PR DESCRIPTION
**Pull request by alonmr**
https://jira.iguazeng.com/browse/ML-4201

### CODE REVIEW

File: mlrun/db/httpdb.py [hunk 1]
```diff
@@ -210,8 +217,14 @@ def api_call(
...
-        if not self.session:
-            self.session = self._init_session()
+        retry_on_post = None
+        if method == "POST":
+            retry_on_post = self._is_retry_on_post_allowed(path)
+
+        # retry_on_post is not None means that method is POST, so we need to update the session
+        # with the appropriate retry policy
+        if not self.session or retry_on_post is not None:
+            self.session = self._init_session(retry_on_post or False)
```
Comment (TomerShor) [inline mlrun/db/httpdb.py:227]: A slightly more readable suggestion
```suggestion
        # if the method is POST, we need to update the session with the appropriate retry policy
           if not self.session or method == "POST":
      retry = self._is_retry_on_post_allowed(path, method)
      self.session = self._init_session(retry)

def _is_retry_on_post_allowed(self, path: str, method):
   if method != "POST":
      return any(re.match(regex, path) for regex in self.RETRIABLE_POST_PATHS)
   return False
```

File: mlrun/db/httpdb.py [hunk 2]
```diff
@@ -98,6 +99,12 @@ class HTTPRunDB(RunDBInterface):
     """
 
     kind = "http"
+    # by default we don't retry on POST request as they are usually not idempotent
+    # here is a list of identified POST requests that are idempotent ('store' vs 'create') and can be retried
+    RETRIABLE_POST_PATHS = [
+        r"\/?projects\/.*\/artifacts\/.*\/.*",
+        r"\/?run\/.*\/.*",
```
Comment (quaark) [inline mlrun/db/httpdb.py:106]: `.*` means 0 or more chars, `.+` is one or more.
```suggestion
        r"\/?projects\/.+\/artifacts\/.+\/.+",
        r"\/?run\/.+\/.+",
```

**Review by quaark — CHANGES_REQUESTED** [review]
minor

**Review by TomerShor — CHANGES_REQUESTED** [review]

**Review by TomerShor — APPROVED** [review]
👌

File: mlrun/db/httpdb.py [hunk 3]
```diff
@@ -239,16 +248,22 @@ def api_call(
...
+            == mlrun.common.schemas.HTTPSessionRetryMode.enabled.value,
+            retry_on_post=retry_on_post,
         )
 
     def _path_of(self, prefix, project, uid):
         project = project or config.default_project
         return f"{prefix}/{project}/{uid}"
 
+    def _is_retry_on_post_allowed(self, method, path: str):
+        return method == "POST" and any(
```
Comment (quaark) [inline mlrun/db/httpdb.py:263]: if you are here then method is definitely POST, don't think you need to check again

**Review by quaark — CHANGES_REQUESTED** [review]

File: mlrun/db/httpdb.py (hunk 3 above)
Comment (alonmr) [reply mlrun/db/httpdb.py:263]: added a comment

**Pull request merged**